### PR TITLE
Feature point inside check for axis aligned geom

### PR DESCRIPTION
**Pull request by Davknapp**
Implemented a point inside check that is specialized for axis-aligned geometries. 

Also enhanced the gtest to test this feature. 

**_All these boxes must be checked by the reviewers before merging the pull request:_**

As a reviewer please read through all the code lines and make sure that the code is fully understood, bug free, well-documented and well-structured.

#### General
- [x] The reviewer executed the new code features at least once and checked the results manually

- [x] The code follows the [t8code coding guidelines](https://github.com/holke/t8code/wiki/Coding-Guideline)
- [x] New source/header files are properly added to the Makefiles
- [x] The code is well documented
- [x] All function declarations, structs/classes and their members have a proper doxygen documentation
- [x] All new algorithms and data structures are sufficiently optimal in terms of memory and runtime (If this should be merged, but there is still potential for optimization, create a new issue)

#### Tests
- [x] The code is covered in an existing or new test case using Google Test

#### Github action

- [x] The code compiles without warning in debugging and release mode, with and without MPI (this should be executed automatically in a github action)
- [x] All tests pass (in various configurations, this should be executed automatically in a github action)

  If the Pull request introduces code that is not covered by the github action (for example coupling with a new library):
  - [ ] Should this use case be added to the github action?
  - [ ] If not, does the specific use case compile and all tests pass (check manually)

#### Scripts and Wiki

- [x] If a new directory with source-files is added, it must be covered by the `script/find_all_source_files.scp` to check the indentation of these files.
- [x] If this PR introduces a new feature, it must be covered in an example/tutorial and a Wiki article.

#### Licence

- [x] The author added a BSD statement to `doc/` (or already has one)

### CODE REVIEW

**Comment by sandro-elsweijer** [comment]
had to do some merging, because the diff showed changes which are not actually from this pr

File: src/t8_forest/t8_forest_cxx.cxx
Comment (sandro-elsweijer) [inline src/t8_forest/t8_forest_cxx.cxx:1246]: For now, we could use [this function](https://github.com/DLR-AMR/t8code/blob/main/src/t8_geometry/t8_geometry_implementations/t8_geometry_linear_axis_aligned.cxx#L82) to check if the geometry is axis aligned. For this, we can remove the debug macro around that function. This way, we get the geometry type directly from the tree and do not have to ask for it via a function argument.
And as soon as we merge PR #731, we can replace this with this [function](https://github.com/DLR-AMR/t8code/blob/feature-puma_patches_stretching/src/t8_geometry/t8_geometry_implementations/t8_geometry_linear_axis_aligned.hxx#L63-L67)

File: src/t8_forest/t8_forest_cxx.cxx
Comment (sandro-elsweijer) [inline src/t8_forest/t8_forest_cxx.cxx:1260]: There is no check, that vertex 0 has all the min coords and vertex 1 has all the max coords. But maybe this is a good requirement, since it allows us to use these assumptions for code optimization

**Review by sandro-elsweijer — CHANGES_REQUESTED** [review]

File: src/t8_forest/t8_forest_cxx.cxx
Comment (Davknapp) [reply src/t8_forest/t8_forest_cxx.cxx]: This check is now performed during cmesh_commit in debug-modus

File: src/t8_forest/t8_forest_cxx.cxx
Comment (Davknapp) [reply src/t8_forest/t8_forest_cxx.cxx]: Then I'll wait for #731 to merge!

File: src/t8_forest/t8_forest_cxx.cxx
Comment (jmark) [inline src/t8_forest/t8_forest_cxx.cxx:1079]: How to calling this function `t8_geom_points_inside_element` ?

File: src/t8_forest/t8_forest_cxx.cxx
Comment (jmark) [inline src/t8_forest/t8_forest_cxx.cxx:1087]: See comment above about the name of this function.

File: src/t8_forest/t8_forest_general.h
Comment (jmark) [inline src/t8_forest/t8_forest_general.h:787]: How about calling this function `t8_forest_element_points_inside` ?

Comment (jmark) [reply src/t8_forest/t8_forest_general.h:787]: I would also consider to remove `t8_forest_element_point_inside` in the midterm future.

File: src/t8_forest/t8_forest_general.h
Comment (jmark) [inline src/t8_forest/t8_forest_general.h:789]: The flag `geom_is_axis_aligned` should not be necessary anymore? Why is it still there?

File: src/t8_geometry/t8_geometry.cxx
Comment (jmark) [inline src/t8_geometry/t8_geometry.cxx:256]: Interesting. Why remove `const` qualifier?

File: src/t8_geometry/t8_geometry_implementations/t8_geometry_examples.hxx
Comment (jmark) [inline src/t8_geometry/t8_geometry_implementations/t8_geometry_examples.hxx:78]: Why repeat this routine here. It derives from `geometry_with_vertices` where this routine is already implemented as virtual function.

File: src/t8_geometry/t8_geometry_implementations/t8_geometry_linear.hxx
Comment (jmark) [inline src/t8_geometry/t8_geometry_implementations/t8_geometry_linear.hxx:93]: Why repeat this routine here. It derives from `geometry_with_vertices` where this routine is already implemented as virtual function.

File: src/t8_geometry/t8_geometry_implementations/t8_geometry_linear_axis_aligned.hxx
Comment (jmark) [inline src/t8_geometry/t8_geometry_implementations/t8_geometry_linear_axis_aligned.hxx:96]: Why repeat this routine here. It derives from `geometry_with_vertices` where this routine is already implemented as virtual function and you also overwrite it in the associated `cxx` file.

File: src/t8_geometry/t8_geometry_implementations/t8_geometry_occ.hxx
Comment (jmark) [inline src/t8_geometry/t8_geometry_implementations/t8_geometry_occ.hxx:121]: Why repeat this routine in `occ` geometry. It derives from `geometry_with_vertices` where this routine is already implemented as virtual function.

File: src/t8_geometry/t8_geometry_implementations/t8_geometry_linear.cxx
Comment (jmark) [inline src/t8_geometry/t8_geometry_implementations/t8_geometry_linear.cxx:1]: I'll go through this file once this PR is in a final state close to merging.

File: src/t8_cmesh/t8_cmesh.c
Comment (jmark) [inline src/t8_cmesh/t8_cmesh.c:543]: Since we are already doing the effort. How about providing this functionality in the geometry module as well? You probably still need `t8_cmesh_tree_vertices_negative_volume` like in `cmesh_readmshfile`. But here we could simply call the specialized geometry member function.

File: example/geometry/t8_example_geometries.cxx
Comment (jmark) [inline example/geometry/t8_example_geometries.cxx:121]: Why not put this routine as virtual function into `t8_geometry`. Then all derived geometry classes have this routine and do not have to put this everywhere.

**Review by jmark — CHANGES_REQUESTED** [review]

File: src/t8_cmesh/t8_cmesh_geometry.cxx
Comment (Davknapp) [reply src/t8_cmesh/t8_cmesh_geometry.cxx:56]: Made the member-function const to readd the const qualifier

File: src/t8_forest/t8_forest_cxx.cxx
Comment (Davknapp) [reply src/t8_forest/t8_forest_cxx.cxx:1102]: I usually don't like to use the plural form, because they can be misleading so easily. the difference between
```t8_geom_points_inside_element``` and 
```t8_geom_point_inside_element``` is not that much

File: src/t8_forest/t8_forest_general.h
Comment (Davknapp) [reply src/t8_forest/t8_forest_general.h:787]: That would be okay for me, to change the name to points, and remove the function for a single point.

File: src/t8_forest/t8_forest_general.h
Comment (Davknapp) [reply src/t8_forest/t8_forest_general.h:789]: Thanks for the catch, removed it.

File: src/t8_geometry/t8_geometry.cxx
Comment (Davknapp) [reply src/t8_geometry/t8_geometry.cxx:256]: I put it back .

File: src/t8_forest/t8_forest_cxx.cxx
Comment (Davknapp) [reply src/t8_forest/t8_forest_cxx.cxx:1102]: Applied your changes after reading your following comments

File: src/t8_cmesh/t8_cmesh.c
Comment (Davknapp) [reply src/t8_cmesh/t8_cmesh.c:543]: Good idea, but I will implement this function on a different branch.

**Review by jmark — DISMISSED** [review]
@Davknapp On my machine I get the hint that this branch is not up-to-date with `main`. Please try a merge with an up-to-date `main` and resolve merge conflicts.

**Comment by jmark** [comment]
> @Davknapp On my machine I get the hint that this branch is not up-to-date with `main`. Please try a merge with an up-to-date `main` and resolve merge conflicts.

Never mind. Forgot to pull latest changes.

**Review by jmark — APPROVED** [review]
LGTM!

File: src/t8_cmesh/t8_cmesh_examples.c
Comment (sandro-elsweijer) [inline src/t8_cmesh/t8_cmesh_examples.c:988]: Isn't this a duplicate of the next two code lines?

File: src/t8_geometry/t8_geometry_implementations/t8_geometry_linear_axis_aligned.cxx
Comment (sandro-elsweijer) [inline src/t8_geometry/t8_geometry_implementations/t8_geometry_linear_axis_aligned.cxx:118]: The new `t8_geom_get_type` function is the replacement for this function. I just removed it from all geometry classes

File: src/t8_cmesh/t8_cmesh_examples.c
Comment (sandro-elsweijer) [inline src/t8_cmesh/t8_cmesh_examples.c:1282]: Why did you change this?

File: src/t8_geometry/t8_geometry_implementations/t8_geometry_examples.hxx
Comment (sandro-elsweijer) [inline src/t8_geometry/t8_geometry_implementations/t8_geometry_examples.hxx:84]: This should crash t8code. The load tree data function is executed every time the tree is loaded into memory

Comment (sandro-elsweijer) [reply src/t8_geometry/t8_geometry_implementations/t8_geometry_examples.hxx:84]: Or it should at least crash the example

**Review by sandro-elsweijer — CHANGES_REQUESTED** [review]
Preliminary review, I am going to eat now

File: src/t8_geometry/t8_geometry_implementations/t8_geometry_examples.hxx
Comment (Davknapp) [reply src/t8_geometry/t8_geometry_implementations/t8_geometry_examples.hxx:84]: This part of the code has not been developed by me. It possibly comes from a merge?

Comment (sandro-elsweijer) [reply src/t8_geometry/t8_geometry_implementations/t8_geometry_examples.hxx:84]: Yes, apparently it was in main once, but after catching the bug it got removed:
https://github.com/DLR-AMR/t8code/commit/00b61eafd93a850e65b394c2136c31b11f3e5def
This is the second thing which tries to re-enter main via this PR. We definitely have to read everything very carefully so that nothing else tries to sneak in.

File: src/t8_cmesh/t8_cmesh_examples.c
Comment (sandro-elsweijer) [inline src/t8_cmesh/t8_cmesh_examples.c:1320]: Is this change intended or an artifact?

File: src/t8_geometry/t8_geometry_base.hxx
Comment (sandro-elsweijer) [inline src/t8_geometry/t8_geometry_base.hxx:121]: Why do we not make this function purely virtual? This way we would have to override it

File: src/t8_geometry/t8_geometry_implementations/t8_geometry_linear.cxx
Comment (sandro-elsweijer) [inline src/t8_geometry/t8_geometry_implementations/t8_geometry_linear.cxx:305]: But wouldn't this also return true if the point is on the line, but not between both vertices?

File: src/t8_geometry/t8_geometry_implementations/t8_geometry_linear.cxx
Comment (sandro-elsweijer) [inline src/t8_geometry/t8_geometry_implementations/t8_geometry_linear.cxx:67]: Maybe we can move these function to the geometry helpers file instead, since other geometries could use them as well

File: src/t8_geometry/t8_geometry_implementations/t8_geometry_zero.hxx
Comment (sandro-elsweijer) [inline src/t8_geometry/t8_geometry_implementations/t8_geometry_zero.hxx:103]:
```suggestion
const int zeros[3] = { 0 }; 
for (int i_point = 0; i_point < num_points; ++i_point)
{
  const int offset = i_point * T8_ECLASS_MAX_DIM;
  is_inside[i_point] = t8_vertex_point_inside(zeros, points + offset, tolerance)
}
```

File: tutorials/general/t8_tutorial_search.cxx
Comment (sandro-elsweijer) [inline tutorials/general/t8_tutorial_search.cxx:376]: Is this meant to be in this PR?

**Review by sandro-elsweijer — CHANGES_REQUESTED** [review]

File: src/t8_geometry/t8_geometry_implementations/t8_geometry_linear.cxx
Comment (Davknapp) [reply src/t8_geometry/t8_geometry_implementations/t8_geometry_linear.cxx:305]: No, because the line check tests if the point can be defined by `p_0 + x * v` with `-eps < x < 1 + eps`

File: tutorials/general/t8_tutorial_search.cxx
Comment (Davknapp) [reply tutorials/general/t8_tutorial_search.cxx:376]: no

File: src/t8_geometry/t8_geometry_implementations/t8_geometry_zero.hxx
Comment (Davknapp) [reply src/t8_geometry/t8_geometry_implementations/t8_geometry_zero.hxx:103]: The zero-geometry does not inherit from `t8_geometry_with_vertices' , therefore I don't think we should implement it this way.

File: src/t8_geometry/t8_geometry_implementations/t8_geometry_linear.cxx
Comment (Davknapp) [reply src/t8_geometry/t8_geometry_implementations/t8_geometry_linear.cxx:67]: Done

File: src/t8_geometry/t8_geometry_base.hxx
Comment (Davknapp) [reply src/t8_geometry/t8_geometry_base.hxx:121]: For a pure virutal function every geometry has to provide such a check. Which can be complex in general.

File: src/t8_cmesh/t8_cmesh_examples.c
Comment (Davknapp) [reply src/t8_cmesh/t8_cmesh_examples.c:1320]: Yes, as the comment above this line says, we want to set the second vertex here, which is the third argument of this function.

File: src/t8_geometry/t8_geometry_base.hxx
Comment (sandro-elsweijer) [reply src/t8_geometry/t8_geometry_base.hxx:121]: ahh okay

File: src/t8_geometry/t8_geometry_implementations/t8_geometry_zero.hxx
Comment (sandro-elsweijer) [reply src/t8_geometry/t8_geometry_implementations/t8_geometry_zero.hxx:103]: Yes it does not inherit from it, but in the end it is still used to compute vertices of the mesh. How would you do it?

**Review by sandro-elsweijer — APPROVED** [review]
One last thing

**Pull request merged**
